compiler/src/dotty/tools/dotc/core/Types.scala

```diff
@@ -3282,7 +3290,12 @@ object Types {
           else tp1.atoms | tp2.atoms
         val tp1w = tp1.widenSingletons
         val tp2w = tp2.widenSingletons
-        myWidened = if ((tp1 eq tp1w) && (tp2 eq tp2w)) this else tp1w | tp2w
+        myWidened =
+          if isSoft then
```

Comment:
With this change, `"a" |<hard> String` would not be widened to `String` anymore. Could add something like `|| tp1w <:< tp2w || tp2w <:< tp1w` here if we want that.